ml_package_framework/data_processing/data_processors.py

```diff
@@ -77,7 +78,262 @@ def load_data(self, path: str, *args, **kwargs) -> pd.DataFrame:
+        It then calculates the total number of orders for each user and a rating feature,
+        which is the ratio of 'ordered' to 'orders_total'. Finally, it creates a unique 'id'
+        for each row, composed of the 'user_id' and 'category' separated by a semicolon.
+
+        Parameters
+        ----------
+        train_melt : pd.DataFrame
+            The melted training dataframe in long format, with 'user_id', 'category',
+            and 'ordered' columns.
+        valid_melt : pd.DataFrame
+            The melted validation dataframe in long format, with at least 'user_id'
+            and 'order_number' columns.
+
+        Returns
+        -------
+        pd.DataFrame
+            The enhanced training dataframe with additional features: 'orders_total',
+            'rating', and 'id'.
+        """
+        train_df = train_melt.copy()
+        order_number = (
+            valid_melt[["user_id", "order_number"]].set_index("user_id").squeeze()
+        )
+        train_df["orders_total"] = train_df["user_id"].map(order_number)
+        train_df["rating"] = train_df["ordered"] / train_df["orders_total"]
+        train_df["id"] = train_df["user_id"].astype(str) + ";" + train_df["category"]
+        return train_df
+
+    def merge_and_filter_(
+        self,
+        train_df: pd.DataFrame,
+        valid_melt: pd.DataFrame,
+        submission_df: pd.DataFrame,
+    ) -> pd.DataFrame:
+        """
+        Merge the training dataset with validation data and filter based on a submission dataframe.
+
+        This method augments the training dataset (train_df) with target values from the validation
+        dataset (valid_melt). It then filters the rows in the training dataset based on the
+        unique 'id' values present in the submission dataframe. The 'target' column from the
+        validation dataset is converted to integer type before merging. The final dataset
+        contains only the entries that have corresponding 'id' values in the submission
+        dataframe.
+
+        Parameters
+        ----------
+        train_df : pd.DataFrame
+            The training dataframe, which includes a unique 'id' for each row.
+        valid_melt : pd.DataFrame
+            The validation dataframe, containing at least the 'target' column.
+        submission_df : pd.DataFrame
+            A dataframe used for filtering, containing unique 'id' values.
+
+        Returns
+        -------
+        pd.DataFrame
+            The filtered and merged training dataframe, now containing 'target' values
+            and rows corresponding to the 'id's in the submission dataframe.
+        """
+        train_df["target"] = valid_melt["target"].astype(int)
+        train_df = train_df[train_df.id.isin(submission_df.id.unique())].reset_index(
+            drop=True
+        )
+        return train_df
+
+    def calculate_total_ordered_(self, train_df: pd.DataFrame) -> pd.DataFrame:
```

Comment:
I dont think you need to use trailing underscore here. It is used to avoid conflicts with existing or built-in functions.